### PR TITLE
More fixes to make Valgrind happy

### PR DESCRIPTION
**Pull request by smharper**
Four fixes here:

In hdf5_interface.h we had prepared buffers for xtensor arrays by using `T* buffer = new T[size]`.  xtensor then deallocates them with (I think) `allocator<double>::deallocate` which causes Valgrind to warn about mismatched free() / delete / delete [].  I've fixed that by using `std::vector` as the buffer.

When we called `read_attribute_string` from Fortran, the buffer was undersized by one byte.  (I assume we miscounted a null character somewhere).  That's fixed now.

When computing photon production cross sections, we were interpolating on a `Tabulated1D` with a value equal to the first grid point in that table.  In that case, our search returned an index of `-1` which we then indexed an array with.

You know all those `may be used uninitialized in this function` warnings we used to get with strings in hdf5_interface.F90?  Valgrind helped me figure out what that was about, and it's now fixed by making sure we allocate character arrays that are declared locally in hdf5_interface.F90 rather than those with inferred length returned by `to_c_string`.

There are still a few small Valgrind errors raised by calls inside the HDF5 library which probably means we've messed up a buffer or a type or something, but I'm not sure now how to track that down.

### CODE REVIEW

File: include/openmc/hdf5_interface.h [hunk 1]
```diff
@@ -162,13 +162,14 @@ void read_attribute(hid_t obj_id, const char* name, xt::xarray<T>& arr)
   std::size_t size = 1;
   for (const auto x : shape)
     size *= x;
-  T* buffer = new T[size];
+  std::vector<T> buffer;
+  buffer.resize(size);
```
Comment (paulromano) [inline include/openmc/hdf5_interface.h]: `std::vector<T> buffer(size)` would be equivalent.

File: src/hdf5_interface.F90 [hunk 2]
```diff
@@ -1320,7 +1331,7 @@ subroutine read_attribute_string(buffer, obj_id, name)
 
     ! Allocate a C char array to get string
     n = attribute_typesize(obj_id, to_c_string(name))
-    allocate(buffer_(n))
+    allocate(buffer_(n+1))
```
Comment (paulromano) [inline src/hdf5_interface.F90]: Yeah, I noticed this one too, but I actually have a better fix for it (and the other HDF5-related errors). Do you mind reverting this change and then I'll put in another PR that fixes the remaining leaks?

**Comment by paulromano** [comment]
I see you removed all the `allocate(name_(len_trim(name) + 1))`. Was there something wrong with it?

**Comment by smharper** [comment]
Oh, whoops.  I thought that's what you meant when you said to remove the other HDF5 errors.  And then I rebased like a fool :p  I'll add those back in.

**Comment by paulromano** [comment]
No, just the one line change where you incremented the size of a character variable by one (I see that one got removed too).

**Pull request merged**